File: objc/source/abstract/WINQ/identifier/Expression.cpp
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <WCDB/WINQ.h>

namespace WCDB {

Expression::Expression(const LiteralValue& literalValue)
{
    syntax().switcher = SyntaxType::Switch::LiteralValue;
    syntax().literalValue = literalValue;
}

Expression::Expression(const BindParameter& bindParameter)
{
    syntax().switcher = SyntaxType::Switch::BindParameter;
    syntax().bindParameter = bindParameter;
}

Expression::Expression(const Column& column)
{
    syntax().switcher = SyntaxType::Switch::Column;
    syntax().column = column;
}

Expression& Expression::table(const String& table)
{
    syntax().table = table;
    return *this;
}

Expression& Expression::schema(const Schema& schema)
{
    syntax().schema = schema;
    return *this;
}

Expression::Expression(const Expressions& expressions)
{
    syntax().switcher = SyntaxType::Switch::Expressions;
    syntax().expressions = expressions;
}

Expression::Expression(const RaiseFunction& raiseFunction)
{
    syntax().switcher = SyntaxType::Switch::RaiseFunction;
    syntax().raiseFunction = raiseFunction;
}

Expression Expression::function(const String& function)
{
    Expression expression;
    expression.syntax().switcher = SyntaxType::Switch::Function;
    expression.syntax().function = function;
    expression.syntax().useWildcard = false;
    return expression;
}

Expression& Expression::invoke()
{
    syntax().useWildcard = false;
    return *this;
}

Expression& Expression::invoke(const Expressions& parameters)
{
    for (const Expression& parameter : parameters) {
        syntax().expressions.push_back(parameter);
    }
    syntax().useWildcard = false;
    return *this;
}

Expression& Expression::invokeAll()
{
    syntax().useWildcard = true;
    return *this;
}

Expression& Expression::distinct()
{
    syntax().distinct = true;
    return *this;
}

Expression Expression::cast(const Expression& expression)
{
    Expression result;
    result.syntax().switcher = SyntaxType::Switch::Cast;
    result.syntax().expressions.push_back(expression);
    return result;
}

Expression& Expression::as(const ColumnType& columnType)
{
    syntax().castType = columnType;
    return *this;
}

Expression Expression::exists(const StatementSelect& select)
{
    Expression expression;
    expression.syntax().switcher = SyntaxType::Switch::Exists;
    expression.syntax().isNot = false;
    expression.syntax().select = select;
    return expression;
}

Expression Expression::notExists(const StatementSelect& select)
{
    Expression expression;
    expression.syntax().switcher = SyntaxType::Switch::Exists;
    expression.syntax().isNot = true;
    expression.syntax().select = select;
    return expression;
}

Expression Expression::case_(const Expression& expression)
{
    Expression result;
    result.syntax().switcher = SyntaxType::Switch::Case;
    result.syntax().hasCase = true;
    result.syntax().expressions.push_back(expression);
    return result;
}

Expression Expression::case_()
{
    Expression result;
    result.syntax().hasCase = false;
    result.syntax().switcher = SyntaxType::Switch::Case;
    return result;
}

Expression& Expression::when(const Expression& expression)
{
    syntax().expressions.push_back(expression);
    return *this;
}

Expression& Expression::then(const Expression& expression)
{
    syntax().expressions.push_back(expression);
    return *this;
}

Expression& Expression::else_(const Expression& expression)
{
    syntax().hasElse = true;
    syntax().expressions.push_back(expression);
    return *this;
}

Expression& Expression::escape(const Expression& operand)
{
    syntax().escape = true;
    syntax().expressions.push_back(operand);
    return *this;
}

Expression Expression::windowFunction(const String& function)
{
    Expression expression;
    expression.syntax().switcher = SyntaxType::Switch::Window;
    expression.syntax().function = function;
    expression.syntax().useWildcard = false;
    return expression;
}

Expression Expression::windowFunction(const String& function, const Expressions& parameter)
{
    Expression expression;
    expression.syntax().switcher = SyntaxType::Switch::Window;
    expression.syntax().function = function;
    expression.syntax().useWildcard = false;
    expression.syntax().expressions = parameter;
    return expression;
}

Expression Expression::windowFunctionAll(const String& function)
{
    Expression expression;
    expression.syntax().switcher = SyntaxType::Switch::Window;
    expression.syntax().function = function;
    expression.syntax().useWildcard = true;
    return expression;
}

Expression& Expression::filter(const Filter& filter)
{
    syntax().useFilter = true;
    syntax().filter = filter;
    return *this;
}

Expression& Expression::over(const WindowDef& windowDef)
{
    syntax().windowDef = windowDef;
    return *this;
}

Expression& Expression::over(const String& window)
{
    syntax().windowName = window;
    return *this;
}

Expression Expression::asExpressionOperand() const
{
    return *this;
}

} // namespace WCDB

File: src/swift/winq/identifier/Expression.swift
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import WCDB_Private

public final class Expression: Identifier<CPPExpression> {

    public convenience init(with literalValueConvertible: LiteralValueConvertible) {
        let literalValue = literalValueConvertible.asLiteralValue()
        let cppExpression = withExtendedLifetime(literalValue) {
            WCDBExpressionCreateWithLiteralValue($0.cppObj)
        }
        self.init(with: cppExpression)
    }

    public convenience init(with bindParameter: BindParameter) {
        let cppExpression = withExtendedLifetime(bindParameter) {
            WCDBExpressionCreateWithBindParameter($0.cppObj)
        }
        self.init(with: cppExpression)
    }

    public convenience init(with columnConvertible: ColumnConvertible) {
        let column = columnConvertible.asColumn()
        let cppExpression = withExtendedLifetime(column) {
            WCDBExpressionCreateWithColumn($0.cppObj)
        }
        self.init(with: cppExpression)
    }

    public convenience init(with statementSelect: StatementSelect) {
        let cppExpression = withExtendedLifetime(statementSelect) {
            WCDBExpressionCreateWithSelection($0.cppObj)
        }
        self.init(with: cppExpression)
    }

    public convenience init(with raiseFunc: RaiseFunction) {
        let cppExpression = withExtendedLifetime(raiseFunc) {
            WCDBExpressionCreateWithRaiseFunction($0.cppObj)
        }
        self.init(with: cppExpression)
    }

    public static func function(_ name: String) -> Expression {
        Expression(with: WCDBExpressionCreateWithFunction(name.cString))
    }

    @discardableResult
    public func schema(_ schemaConvertible: SchemaConvertible) -> Expression {
        let schema = schemaConvertible.asSchema()
        withExtendedLifetime(schema) {
            WCDBExpressionSetWithSchema(cppObj, schema.cppObj)
        }
        return self
    }

    @discardableResult
    public func distinct() -> Expression {
        WCDBExpressionDistinct(cppObj)
        return self
    }

    @discardableResult
    public func arguments(_ expressionConvertibleList: [ExpressionConvertible]) -> Expression {
        let expressions = expressionConvertibleList.map { $0.asExpression() }
        withExtendedLifetime(expressions) {
            let arguments = $0.map { $0.cppObj }
            arguments.withUnsafeBufferPointer { cppArgs in
                WCDBExpressionSetArguments(cppObj, cppArgs.baseAddress, Int32(cppArgs.count))
            }
        }
        return self
    }

    @discardableResult
    public func arguments(_ expressionConvertibleList: ExpressionConvertible...) -> Expression {
        return self.arguments(expressionConvertibleList)
    }

    @discardableResult
    public func invoke() -> Expression {
        WCDBExpressionInvoke(cppObj)
        return self
    }

    @discardableResult
    public func invokeAll() -> Expression {
        WCDBExpressionInvokeAll(cppObj)
        return self
    }

    public static func cast(_ expressionConvertible: ExpressionConvertible) -> Expression {
        let expression = expressionConvertible.asExpression()
        let cppExpression = withExtendedLifetime(expression) {
            WCDBExpressionCast($0.cppObj)
        }
        return Expression(with: cppExpression)
    }

    @discardableResult
    public func `as`(_ type: ColumnType) -> Expression {
        WCDBExpressionAs(cppObj, type.cValue)
        return self
    }

    public func `as`(_ alias: String) -> ResultColumn {
        return ResultColumn(with: WCDBExpressionConfigAlias(cppObj, alias.cString))
    }

    public static func `case`(_ expressionConvertible: ExpressionConvertible? = nil) -> Expression {
        if let expressionConvertible = expressionConvertible {
            let expression = expressionConvertible.asExpression()
            let cppExpression = withExtendedLifetime(expression) {
                WCDBExpressionCaseWithExp($0.cppObj)
            }
            return Expression(with: cppExpression)
        } else {
            return Expression(with: WCDBExpressionCase())
        }
    }

    @discardableResult
    public func when(_ expressionConvertible: ExpressionConvertible) -> Expression {
        let expression = expressionConvertible.asExpression()
        withExtendedLifetime(expression) {
            WCDBExpressionSetWithWhenExp(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func then(_ expressionConvertible: ExpressionConvertible) -> Expression {
        let expression = expressionConvertible.asExpression()
        withExtendedLifetime(expression) {
            WCDBExpressionSetWithThenExp(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func `else`(_ expressionConvertible: ExpressionConvertible) -> Expression {
        let expression = expressionConvertible.asExpression()
        withExtendedLifetime(expression) {
            WCDBExpressionSetWithElseExp(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func escape(_ expressionConvertible: ExpressionConvertible) -> Expression {
        let expression = expressionConvertible.asExpression()
        withExtendedLifetime(expression) {
            WCDBExpressionEscapeWith(cppObj, $0.cppObj)
        }
        return self
    }

    public static func windowFunction(_ funcName: String) -> Expression {
        return Expression(with: WCDBExpressionCreateWithWindowFunction(funcName.cString))
    }

    @discardableResult
    public func filter(_ expressionConvertible: ExpressionConvertible) -> Expression {
        let expression = expressionConvertible.asExpression()
        withExtendedLifetime(expression) {
            WCDBExpressionFilter(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func over(_ window: WindowDef) -> Expression {
        withExtendedLifetime(window) {
            WCDBExpressionOverWindowDef(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func over(_ window: String) -> Expression {
        WCDBExpressionOverWindow(cppObj, window.cString)
        return self
    }
}

extension Expression: ExpressibleByNilLiteral {
    public convenience init(nilLiteral value: ()) {
        self.init(with: LiteralValue(value))
    }
}
extension Expression: ExpressibleByBooleanLiteral {
    public convenience init(booleanLiteral value: Bool) {
        self.init(with: LiteralValue(value))
    }
}
extension Expression: ExpressibleByIntegerLiteral {
    public convenience init(integerLiteral value: Int) {
        self.init(with: LiteralValue(value))
    }
}
extension Expression: ExpressibleByFloatLiteral {
    public convenience init(floatLiteral value: Double) {
        self.init(with: LiteralValue(value))
    }
}
extension Expression: ExpressibleByStringLiteral {
    public convenience init(stringLiteral value: String) {
        self.init(with: LiteralValue(value))
    }
}

extension Expression: ExpressionOperable {

    public static func exists(_ select: StatementSelect) -> Expression {
        let cppExpression = withExtendedLifetime(select) {
            WCDBExpressionCreateWithExistStatement($0.cppObj)
        }
        return Expression(with: cppExpression)
    }

    public static func notExists(_ select: StatementSelect) -> Expression {
        let cppExpression = withExtendedLifetime(select) {
            WCDBExpressionCreateWithNotExistStatement($0.cppObj)
        }
        return Expression(with: cppExpression)
    }

    // Function
    static func function(named name: String,
                         _ arguments: ExpressionConvertible...,
                         isDistinct: Bool = false) -> Expression {
        return function(named: name, arguments, isDistinct: isDistinct)
    }
    static func function(named name: String,
                         _ arguments: [ExpressionConvertible],
                         isDistinct: Bool = false) -> Expression {
        let expression = Expression.function(name)
        expression.invoke()
        if isDistinct {
            expression.distinct()
        }
        if arguments.count > 0 {
            expression.arguments(arguments)
        }
        return expression
    }

    public static func `case`(_ expressionConvertible: ExpressionConvertible,
                              _ flows: (when: ExpressionConvertible, then: ExpressionConvertible)...,
                              `else`: ExpressionConvertible) -> Expression {
        return `case`(expressionConvertible, flows, else: `else`)
    }
    public static func `case`(_ `case`: ExpressionConvertible,
                              _ flows: [(when: ExpressionConvertible, then: ExpressionConvertible)],
                              `else`: ExpressionConvertible) -> Expression {
        let expression = Expression.case(`case`)
        for flow in flows {
            expression.when(flow.when).then(flow.then)
        }
        expression.else(`else`)
        return expression
    }

    public func asExpression() -> Expression {
        return self
    }
}
